### PR TITLE
Add langoustine to the existing list of schedulers.

### PR DESCRIPTION
**Pull request by MaximeKestemont**
This adds langoustine to the list of schedulers already integrated to Dr.Elephant. It will most notably allow the "flow history" and "job history" view to work with jobs scheduled with langoustine.

As Langoustine concepts are a bit different than the other schedulers, some changes were needed and it was not as simple as simply adding a new scheduler class. More particularly, we do not have a concept of "flow" or "flow execution id" in Langoustine. The work-around here is to consider the flow as the langoustine project itself (e.g. datamart, product-rollout, etc.), while the flow execution id is the langoustine start date, i.e. the langoustine functional date (which should always correspond to a specific rounded hour, e.g. Jun 19, 2017 04:00 PM).

However, Langoustine currently does not log the project name, and the tempory fix is to use the scheduler configuration file, to map the user name to a specific project name. Users not mapped to a specific project are currently mapped to "all", so that the related jobs can still be viewed in the flow/job history views.

Finally, there is still some integration to do, in a subsequent PR, as some urls are not refered yet. This means that some links between the different views are not working properly, however, there is no regression compared to the previous status, as the views were not even available for langoustine jobs.

### CODE REVIEW

**Comment by AnthonyTruchet** [comment]
You might want to ask specifically @superbobry for a review or for who you should ask a review to :-)

**Comment by superbobry** [comment]
Sadly, I'm not familiar with Langoustine at all, so I think anyone from the Langoustine devs would be better than me.

I'd have a look anyway, it's just that I don't feel confident enough.

File: app/com/linkedin/drelephant/schedulers/LangoustineScheduler.java [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+
+  @Override
+  public String getSchedulerName() {
+    return schedulerName;
+  }
+
+  @Override
+  public boolean isEmpty() {
+    boolean result = jobId == null || jobDate == null || jobWorkflow == null;
+    return result;
```
Comment (superbobry) [inline app/com/linkedin/drelephant/schedulers/LangoustineScheduler.java]: Inline `result`.

File: app/com/linkedin/drelephant/schedulers/LangoustineScheduler.java [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+ */
+public class LangoustineScheduler implements Scheduler {
+
+  private static final Logger logger = Logger.getLogger(LangoustineScheduler.class);
+
+  public static final String LANGOUSTINE_JOB_ID = "com.criteo.langoustine.name";
+  public static final String LANGOUSTINE_START_DATE = "com.criteo.langoustine.calc_start";
+
+  private String schedulerName;
+  private String jobId;
```
Comment (superbobry) [inline app/com/linkedin/drelephant/schedulers/LangoustineScheduler.java]: Consider annotating as `@Nullable`.

File: app/com/linkedin/drelephant/util/InfoExtractor.java [hunk 3]
```diff
@@ -116,6 +116,13 @@ public static void loadInfo(AppResult result, HadoopApplicationData data) {
     } else if ( data instanceof SparkApplicationData) {
       properties = retrieveSparkProperties((SparkApplicationData) data);
     }
+
+    // TODO temporary fix, it should be automatically set by langoustine in the /conf properties
```
Comment (superbobry) [inline app/com/linkedin/drelephant/util/InfoExtractor.java]: Do you plan to address this in Langoustine soon?

File: app/controllers/Application.java [hunk 4]
```diff
@@ -1137,14 +1138,26 @@ public static Result restFlowGraphData(String flowDefId) {
...
+      }
+      dataset.addProperty("flowtime", maybeParsed);
       dataset.addProperty("score", flowPerfScore);
       dataset.add("jobscores", jobScores);
 
       datasets.add(dataset);
     }
 
-    return ok(new Gson().toJson(datasets));
+    JsonArray sortedDatasets = sortJsonArray(datasets);
```
Comment (superbobry) [inline app/controllers/Application.java:1149]: Why do you need to sort?

File: app/controllers/Application.java [hunk 5]
```diff
@@ -1137,14 +1138,26 @@ public static Result restFlowGraphData(String flowDefId) {
 
       // Execution record
       JsonObject dataset = new JsonObject();
-      dataset.addProperty("flowtime", mrJobsList.get(mrJobsList.size() - 1).finishTime);
+
+      // TODO should make it specific to langoustine as other scheduler are not storing the date in flowExecId field
```
Comment (superbobry) [inline app/controllers/Application.java]: Could you elaborate just to set the context here? How do the other schedulers get the date?

File: app/controllers/Application.java [hunk 6]
```diff
@@ -1228,14 +1241,26 @@ public static Result restJobGraphData(String jobDefId) {
...
```
Comment (superbobry) [inline app/controllers/Application.java]: Copy/paste.

File: app/controllers/Application.java [hunk 7]
```diff
@@ -1228,14 +1241,26 @@ public static Result restJobGraphData(String jobDefId) {
...
+      }
+      dataset.addProperty("flowtime", maybeParsed);
       dataset.addProperty("score", jobPerfScore);
       dataset.add("stagescores", stageScores);
 
       datasets.add(dataset);
     }
 
-    return ok(new Gson().toJson(datasets));
+    JsonArray sortedDatasets = sortJsonArray(datasets);
```
Comment (superbobry) [inline app/controllers/Application.java:1244]: Same question as above: why change the behaviour and sort?

File: app/controllers/Application.java [hunk 8]
```diff
@@ -1330,7 +1355,17 @@ public static Result restJobMetricsGraphData(String jobDefId) {
 
       // Execution record
       JsonObject dataset = new JsonObject();
-      dataset.addProperty("flowtime", mrJobsList.get(mrJobsList.size() - 1).finishTime);
+
+      // TODO should make it specific to langoustine as other scheduler are not storing the date in flowExecId field
```
Comment (superbobry) [inline app/controllers/Application.java]: Copy/paste.

File: app/controllers/Application.java [hunk 9]
```diff
@@ -1480,7 +1517,17 @@ public static Result restFlowMetricsGraphData(String flowDefId) {
...
```
Comment (superbobry) [inline app/controllers/Application.java]: Copy/paste.

File: app/controllers/Application.java (hunk 4 above)
Comment (superbobry) [reply app/controllers/Application.java:1149]: Also: can we sort before converting to JSON?

File: build.sbt [hunk 10]
```diff
@@ -25,6 +25,9 @@ organization := "com.linkedin.drelephant"
 
 javacOptions in Compile ++= Seq("-source", "1.6", "-target", "1.6")
 
+// Prevent failure on some environment
```
Comment (superbobry) [inline build.sbt:28]: Could you be more specific? 

It seems this can be fixed directly in the environment by setting `LANG` to the appropriate locale string.

File: test/com/linkedin/drelephant/util/InfoExtractorTest.java [hunk 11]
```diff
@@ -218,14 +218,14 @@ public boolean isEmpty() {
...
     assertEquals(result.scheduler, "azkaban");
-    assertFalse(StringUtils.isEmpty(result.getJobExecId()));
-    assertFalse(StringUtils.isEmpty(result.getJobDefId()));
-    assertFalse(StringUtils.isEmpty(result.getFlowExecId()));
-    assertFalse(StringUtils.isEmpty(result.getFlowDefId()));
-    assertFalse(StringUtils.isEmpty(result.getJobExecUrl()));
-    assertFalse(StringUtils.isEmpty(result.getJobDefUrl()));
-    assertFalse(StringUtils.isEmpty(result.getFlowExecUrl()));
-    assertFalse(StringUtils.isEmpty(result.getFlowDefUrl()));
+    assertFalse(StringUtils.isEmpty(result.jobExecId));
```
Comment (superbobry) [inline test/com/linkedin/drelephant/util/InfoExtractorTest.java]: Seems unrelated to Langoustine, but in any case, why not the getters?

File: app/controllers/Application.java [hunk 12]
```diff
@@ -123,6 +123,7 @@
   private static int _numJobsCritical = 0;
   private static int _numJobsSevere = 0;
 
+  private static SimpleDateFormat sdf = new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss");
```
Comment (superbobry) [inline app/controllers/Application.java]: Should be UPPER_SNAKE_CASE.

**Review by superbobry — COMMENTED** [review]
I think this is too hacky to merge just yet, unless you want to add basic support for Langoustine ASAP.

Personally, I'd prefer to have a complete solution in one PR instead of two PRs with some ad-hoc code in between. However, this is not a strong opinion and I'm not one of Langoustine users.

File: app/controllers/Application.java (hunk 5 above)
Comment (MaximeKestemont) [reply app/controllers/Application.java]: Most schedulers are based on the concept of flow (which is a workflow of jobs), and of specific execution of those flows. For example, you have test_workflow, and each day, you execute that specific workflow, with an associated flow_execution_id (hashed from the date for example).

In Langoustine, you do not have those 2 concepts directly. The workflow can be identified as being the whole langoustine project (e.g. Sales-matching), so this one is easy.
The flow_execution_id though is a bit different. We do not have any uid that could match a specific flow execution (in fact, we do not even have specific flow execution, it is more "streamlined" with the concept of time series). So the work-around I am using here is to consider that the flow_execution_id is in fact the langoustine_start_date, which correspond to the functional start date of the job.

Does it clarify a bit the context ?

File: app/controllers/Application.java (hunk 12 above)
Comment (MaximeKestemont) [reply app/controllers/Application.java]: Done.

File: app/com/linkedin/drelephant/util/InfoExtractor.java (hunk 3 above)
Comment (MaximeKestemont) [reply app/com/linkedin/drelephant/util/InfoExtractor.java]: I am planning to address it yes, either in this PR or in a following one. There is a current MR on langoustine to add the project name in the /conf properties, that will then be retrieved through the job history api.

File: app/com/linkedin/drelephant/schedulers/LangoustineScheduler.java (hunk 1 above)
Comment (MaximeKestemont) [reply app/com/linkedin/drelephant/schedulers/LangoustineScheduler.java]: Done.

File: app/com/linkedin/drelephant/schedulers/LangoustineScheduler.java (hunk 2 above)
Comment (MaximeKestemont) [reply app/com/linkedin/drelephant/schedulers/LangoustineScheduler.java]: This is not done in the other schedulers classes (AzkabanScheduler for example, which is the Linkedin scheduler), so I was following the same trend as to keep homogeneity.

File: app/controllers/Application.java (hunk 4 above)
Comment (MaximeKestemont) [reply app/controllers/Application.java:1149]: The front-end is displaying a table with the execution_id, in the Flow History and Job History views, and there is no ordering done there. It is based on the fact that the ordering is already done in the back-end, which was the case before as it was ordered based on the job_finish_date (which is the order in which the data are fetched from the job history). Which btw was a bit unintuitive as you are thus not able to view the execution_id in this view, but this is another issue (that is solved for Langoustine in our case).

As I am sightly changing the way this view is working, so that the key in the table is the execution_id (which in our case, is the langoustine functional start date), this is not naturally ordered anymore, as a job that is executed later on can have a functional date that is before another job (that was executed earlier).


And for the second question, I guess we could, but that means a bit of refactoring, as it is working with JsonArray from the start. So imo, this is out of the scope of this PR.

File: app/controllers/Application.java (hunk 6 above)
Comment (MaximeKestemont) [reply app/controllers/Application.java]: I removed the copy/paste and also added an additional check to make it specific to langoustine.

File: build.sbt (hunk 10 above)
Comment (MaximeKestemont) [reply build.sbt:28]: More specifically, 7 tests fails with the error : 
`[error] Test com.linkedin.drelephant.util.MemoryFormatUtilsTest.testBytesToString failed: expected:<1[,]000 GB> but was:<1[ ]000 GB>`
on a laptop with LANG=fr_FR.UTF-8.

So indeed, changing the LANG env variable should solve the issue, but this require the user to change his settings, which he may not necessarily want. Is my solution not cleaner ? There may be something I am missing here, not really my comfort zone.

File: test/com/linkedin/drelephant/util/InfoExtractorTest.java (hunk 11 above)
Comment (MaximeKestemont) [reply test/com/linkedin/drelephant/util/InfoExtractorTest.java]: Mistake of mine. The getters are not explicitly defined, but I guess there is some macros somewhere that generate the getters. I removed my changes.

**Review by MaximeKestemont — COMMENTED** [review]
I have added correction for most of your comments, thanks for those !

There is still 2 notable things to add : 
- using the project name instead of the user name of langoustine. I am waiting for a change in langoustine for that, but it will probably take several days.
- making the url link works for langoustine, so that you can quickly jump between the different views in the UI. Not sure this is easy to do, I still need to investigate on how to retrieve those urls.

File: app/com/linkedin/drelephant/util/InfoExtractor.java (hunk 3 above)
Comment (MaximeKestemont) [reply app/com/linkedin/drelephant/util/InfoExtractor.java]: This has been removed, as I am now retrieving the project name in a non hacky way, from the /conf properties.This should currently works for all Yarn/Hive job scheduled by Langoustine, as long as the project name is properly defined in the configuration of the project (reference.conf).

File: app-conf/SchedulerConf.xml [hunk 13]
```diff
@@ -39,6 +39,13 @@
     </scheduler>
 
     <scheduler>
+        <name>langoustine</name>
+        <classname>com.linkedin.drelephant.schedulers.LangoustineScheduler</classname>
```
Comment (remysaissy) [inline app-conf/SchedulerConf.xml:43]: Langoustine is not open sourced yet.
It would be a better thing to put it in a separate namespace, as an optional Scheduler to Dr-Elephant.

File: app/controllers/Application.java [hunk 14]
```diff
@@ -1135,18 +1136,23 @@ public static Result restFlowGraphData(String flowDefId) {
         flowPerfScore += jobPerfScore;
       }
 
+
       // Execution record
       JsonObject dataset = new JsonObject();
-      dataset.addProperty("flowtime", mrJobsList.get(mrJobsList.size() - 1).finishTime);
+      dataset.addProperty("flowtime", getFlowTime(mrJobsList.get(mrJobsList.size() - 1)));
```
Comment (remysaissy) [inline app/controllers/Application.java:1142]: Shouldn't it be put in a separate commit as it looks more like a refactoring?

File: app/controllers/Application.java (hunk 4 above)
Comment (remysaissy) [reply app/controllers/Application.java:1149]: You are not breaking any existing behaviour with other scheduler by doing this way?

File: app/controllers/Application.java [hunk 15]
```diff
@@ -1490,7 +1500,61 @@ public static Result restFlowMetricsGraphData(String flowDefId) {
       datasets.add(dataset);
     }
 
-    return ok(new Gson().toJson(datasets));
+    JsonArray sortedDatasets = sortJsonArray(datasets);
```
Comment (remysaissy) [inline app/controllers/Application.java:1503]: This refactoring looks good but I would put it separately from adding Langoustine support to Dr-Elephant.

**Review by remysaissy — CHANGES_REQUESTED** [review]

File: app/controllers/Application.java (hunk 14 above)
Comment (MaximeKestemont) [reply app/controllers/Application.java:1142]: This is a change quite specific to Langoustine in fact. I am not changing anything for other schedulers as the getFlowTime method is keeping the old behaviour for jobs not scheduled with Langoustine. Imo this is part of the Langoustine integration, and thus under the scope of this PR, but it could indeed be moved to another PR, as it could work without this change.
